Mathlib/RingTheory/PowerSeries/Substitution.lean

```diff
@@ -0,0 +1,381 @@
+
+theorem subst_smul [Algebra A S] [IsScalarTower A R S]
+    (ha : HasSubst a) (r : A) (f : PowerSeries R) :
+    subst a (r • f) = r • (subst a f) := by
+  rw [← coe_substAlgHom ha, AlgHom.map_smul_of_tower]
+
+theorem coeff_subst_finite (ha : HasSubst a) (f : PowerSeries R) (e : τ →₀ ℕ) :
+    Set.Finite (fun (d : ℕ) ↦ (coeff R d f) • (MvPowerSeries.coeff S e (a ^ d))).support := by
+  convert (MvPowerSeries.coeff_subst_finite ha.const f e).image
+    ⇑(Finsupp.LinearEquiv.finsuppUnique ℕ ℕ Unit).toEquiv
+  rw [← Equiv.preimage_eq_iff_eq_image, ← Function.support_comp_eq_preimage]
+  apply congr_arg
+  rw [← Equiv.eq_comp_symm]
+  ext d
+  simp only [Finsupp.prod_pow, Finset.univ_unique, PUnit.default_eq_unit, Finset.prod_singleton,
+    LinearEquiv.coe_toEquiv_symm, EquivLike.coe_coe, Function.comp_apply,
+    Finsupp.LinearEquiv.finsuppUnique_symm_apply, Finsupp.single_eq_same]
+  congr
+
+theorem coeff_subst_finite' (hb : HasSubst b) (f : PowerSeries R) (e : ℕ) :
+    Set.Finite (fun (d : ℕ) ↦ (coeff R d f) • (PowerSeries.coeff S e (b ^ d))).support :=
+  coeff_subst_finite hb f  _
+
+theorem coeff_subst (ha : HasSubst a) (f : PowerSeries R) (e : τ →₀ ℕ) :
+    MvPowerSeries.coeff S e (subst a f) =
+      finsum (fun (d : ℕ) ↦
+        (coeff R d f) • (MvPowerSeries.coeff S e (a ^ d))) := by
+  erw [MvPowerSeries.coeff_subst ha.const f e]
+  rw [← finsum_comp_equiv (Finsupp.LinearEquiv.finsuppUnique ℕ ℕ Unit).toEquiv.symm]
+  apply finsum_congr
+  intro d
+  congr
+  · ext; simp
+  · simp
+
+theorem coeff_subst' {b : S⟦X⟧} (hb : HasSubst b) (f : R⟦X⟧) (e : ℕ) :
+    coeff S e (f.subst b) =
+      finsum (fun (d : ℕ) ↦
+        (coeff R d f) • (PowerSeries.coeff S e (b ^ d))) := by
+  simp only [PowerSeries.coeff, coeff_subst hb]
+
+theorem constantCoeff_subst (ha : HasSubst a) (f : PowerSeries R) :
+    MvPowerSeries.constantCoeff τ S (subst a f) =
+      finsum (fun d ↦ (coeff R d f) • (MvPowerSeries.constantCoeff τ S (a ^ d))) := by
+  simp only [← MvPowerSeries.coeff_zero_eq_constantCoeff_apply, coeff_subst ha f 0]
+
+theorem map_algebraMap_eq_subst_X (f : R⟦X⟧) :
+    map (algebraMap R S) f = subst X f :=
+  MvPowerSeries.map_algebraMap_eq_subst_X f
+
+theorem _root_.Polynomial.toPowerSeries_toMvPowerSeries (p : Polynomial R) :
+    (p : PowerSeries R) =
+      ((Polynomial.aeval (MvPolynomial.X ()) p : MvPolynomial Unit R) : MvPowerSeries Unit R) := by
+  change (Polynomial.coeToPowerSeries.algHom R) p =
+    (MvPolynomial.coeToMvPowerSeries.algHom R)
+      (Polynomial.aeval (MvPolynomial.X () : MvPolynomial Unit R) p)
+  rw [← AlgHom.comp_apply]
+  apply AlgHom.congr_fun
+  apply Polynomial.algHom_ext
+  simp only [Polynomial.coeToPowerSeries.algHom_apply, Algebra.id.map_eq_id, Polynomial.coe_X,
+    map_X]
+  erw [AlgHom.comp_apply]
+  simp only [Polynomial.aeval_X, MvPolynomial.coeToMvPowerSeries.algHom_apply,
+    Algebra.id.map_eq_id, MvPowerSeries.map_id, MvPolynomial.coe_X, RingHom.id_apply]
+  rfl
```

Comment:
Here the goal is `X = MvPowerSeries.X ()`. Can you add a `rfl` lemma saying that and rewriting it here (the lemma can be called `X_eq'` and go after [PowerSeries.X_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/PowerSeries/Basic.html#PowerSeries.X_eq))?